### PR TITLE
lua: align the allocated memory if required

### LINKED ISSUE #5551: lua, osx: envoy with lua http HTTP filter enabled crashes on release build
*Title*: Lua: envoy with Lua HTTP filter enable crashes on release build on osx

*Description*:
Envoy with Lua HTTP filter crashes on release build. Seems like something is wrong (failed to do the placement to the allocated mem?) when executing the placement new operator [here](https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/common/lua/lua.h#L101) for the `StreamHandleWrapper`.

```cpp
 ENVOY_LOG(trace, "creating {} at {}", typeid(T).name(), mem);
 return {new (mem) T(std::forward<ConstructorArgs>(args)...), state};
```

*Repro steps*:

Build Envoy using `-c opt`.

```
$ bazel --bazelrc=/dev/null build -c opt //source/exe:envoy-static.stripped
$ ./bazel-bin/source/exe/envoy-static.stripped -c config.yaml
$ curl localhost:10000
// 10000 is the listener port
```

Where an example of config.yaml is as follows:

```yaml
static_resources:
  listeners:
  - name: main
    address:
      socket_address:
        address: 0.0.0.0
        port_value: 10000
    filter_chains:
    - filters:
      - name: envoy.http_connection_manager
        config:
          stat_prefix: ingress_http
          route_config:
            name: local_route
            virtual_hosts:
            - name: local_service
              domains: ["*"]
              routes:
              - match:
                  prefix: "/"
                route:
                  host_rewrite: www.google.com
                  cluster: service_google
          http_filters:
          - name: envoy.lua
            config:
              inline_code: |
                function envoy_on_request(request_handle)
                end
                function envoy_on_response(response_handle)
                end
          - name: envoy.router
            config: {}

  clusters:
  - name: service_google
    connect_timeout: 0.25s
    type: LOGICAL_DNS
    # Comment out the following line to test on v6 networks
    dns_lookup_family: V4_ONLY
    lb_policy: ROUND_ROBIN
    hosts:
      - socket_address:
          address: google.com
          port_value: 443
    tls_context: { sni: www.google.com }
admin:
  access_log_path: "/dev/null"
  address:
    socket_address:
      address: 0.0.0.0
      port_value: 5000
```

*Logs*:

```
[2019-01-09 19:13:23.818][4252367][debug][http] [source/common/http/conn_manager_impl.cc:539] [C0][S4589606042074743932] request headers complete (end_stream=true):
':authority', 'localhost:10000'
':path', '/'
':method', 'GET'
'user-agent', 'curl/7.62.0'
'accept', '*/*'

[2019-01-09 19:13:23.818][4252367][debug][http] [source/common/http/conn_manager_impl.cc:988] [C0][S4589606042074743932] request end stream
[2019-01-09 19:13:23.818][4252367][trace][lua] [bazel-out/darwin-opt/bin/source/extensions/filters/common/lua/_virtual_includes/lua_lib/extensions/filters/common/lua/lua.h:100] creating N5Envoy10Extensions11HttpFilters3Lua19StreamHandleWrapperE at 0x1e0e6a8
[2019-01-09 19:13:23.818][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:125] Caught Segmentation fault: 11, suspect faulting address 0x0
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:89] Backtrace thr<123145554665472> obj<envoy-static.stripped>:
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:104] thr<123145554665472> obj<envoy-static.stripped               0x0000000100f0bbf5 _ZN5Envoy10Extensions7Filters6Common3Lua13BaseLuaObjectINS0_11HttpFilters3Lua19StreamHandleWrapperEE6createIJRNS3_9CoroutineERNS_4Http9HeaderMapERbRNS6_6FilterERNS6_15FilterCallbacksEEEENSt3__14pairIPS7_P9lua_StateEESO_DpOT_>
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:110] thr<123145554665472> #0 0x100f0bbf5
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:104] thr<123145554665472> obj<envoy-static.stripped>
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:110] thr<123145554665472> #1 0x100f0b9f1 Envoy::Extensions::HttpFilters::Lua::Filter::doHeaders(Envoy::Extensions::Filters::Common::Lua::LuaDeathRef<Envoy::Extensions::HttpFilters::Lua::StreamHandleWrapper>&, std::__1::unique_ptr<Envoy::Extensions::Filters::Common::Lua::Coroutine, std::__1::default_delete<Envoy::Extensions::Filters::Common::Lua::Coroutine> >&, Envoy::Extensions::HttpFilters::Lua::FilterCallbacks&, int, Envoy::Http::HeaderMap&, bool) + 209
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:110] thr<123145554665472> #2 0x100f0c59c Envoy::Extensions::HttpFilters::Lua::Filter::decodeHeaders(Envoy::Http::HeaderMap&, bool) + 76
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:110] thr<123145554665472> #3 0x1013aff7d Envoy::Http::ConnectionManagerImpl::ActiveStream::decodeHeaders(Envoy::Http::ConnectionManagerImpl::ActiveStreamDecoderFilter*, Envoy::Http::HeaderMap&, bool) + 349
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:110] thr<123145554665472> #4 0x1013af441 Envoy::Http::ConnectionManagerImpl::ActiveStream::decodeHeaders(std::__1::unique_ptr<Envoy::Http::HeaderMap, std::__1::default_delete<Envoy::Http::HeaderMap> >&&, bool) + 2465
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:110] thr<123145554665472> #5 0x101510502 Envoy::Http::Http1::ServerConnectionImpl::onMessageComplete() + 82
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:110] thr<123145554665472> #6 0x10150f522 Envoy::Http::Http1::ConnectionImpl::onMessageCompleteBase() + 226
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:110] thr<123145554665472> #7 0x1015121bc Envoy::Http::Http1::ConnectionImpl::$_7::__invoke(http_parser*) + 12
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:110] thr<123145554665472> #8 0x101514d6f http_parser_execute + 4847
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:110] thr<123145554665472> #9 0x10150efff Envoy::Http::Http1::ConnectionImpl::dispatchSlice(char const*, unsigned long) + 47
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:110] thr<123145554665472> #10 0x10150ee1e Envoy::Http::Http1::ConnectionImpl::dispatch(Envoy::Buffer::Instance&) + 286
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:110] thr<123145554665472> #11 0x1013ac3e8 Envoy::Http::ConnectionManagerImpl::onData(Envoy::Buffer::Instance&, bool) + 264
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:110] thr<123145554665472> #12 0x1011c12e2 Envoy::Network::FilterManagerImpl::onRead() + 130
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:110] thr<123145554665472> #13 0x1011bd986 Envoy::Network::ConnectionImpl::onReadReady() + 358
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:110] thr<123145554665472> #14 0x1011bd524 Envoy::Network::ConnectionImpl::onFileEvent(unsigned int) + 132
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:110] thr<123145554665472> #15 0x1011b76d2 Envoy::Event::FileEventImpl::assignEvents(unsigned int)::$_0::__invoke(int, short, void*) + 50
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:110] thr<123145554665472> #16 0x1015e2086 event_process_active_single_queue + 1766
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:110] thr<123145554665472> #17 0x1015de7cb event_base_loop + 1691
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:110] thr<123145554665472> #18 0x1011b6ae4 Envoy::Event::DispatcherImpl::run(Envoy::Event::Dispatcher::RunType) + 132
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:110] thr<123145554665472> #19 0x1011b159f Envoy::Server::WorkerImpl::threadRoutine(Envoy::Server::GuardDog&) + 191
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:110] thr<123145554665472> #20 0x1018c6c92 Envoy::Thread::ThreadImplPosix::ThreadImplPosix(std::__1::function<void ()>)::$_0::__invoke(void*) + 18
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:104] thr<123145554665472> obj<libsystem_pthread.dylib>
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:110] thr<123145554665472> #21 0x7fff62beb304 _pthread_body + 125
[2019-01-09 19:13:23.821][4252367][critical][backtrace] [bazel-out/darwin-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:121] end backtrace thread 123145554665472
Segmentation fault: 11
```

### PR DESCRIPTION
**Pull request by dio**
*Description*:
This patch aligns the memory that is allocated by Lua. Previously, without alignment, Envoy experienced segfault when constructing `StreamWrapperHandler` using the placement new operator on the pre-allocated (via `new_luauserdata` API) memory by Lua.

*Risk Level*: Medium
*Testing*: Build on mac, check with ASAN, manual testing.
*Docs Changes*: N/A
*Release Notes*: N/A
Fixes #5551 

Signed-off-by: Dhi Aurrahman <dio@tetrate.io>

### CODE REVIEW

**Comment by jmarantz** [comment]
I was trying to see if anyone other than Matt had detailed knowledge our Lua stack. Anyone else want to do a pass before Matt looks at it?

File: source/extensions/filters/common/lua/lua.h [hunk 1]
```diff
@@ -60,6 +60,28 @@ namespace Lua {
  */
 #define DECLARE_LUA_CLOSURE(Class, Name) DECLARE_LUA_FUNCTION_EX(Class, Name, lua_upvalueindex(1))
 
+/**
+ * Calculate the maximum space needed to be aligned.
+ */
+template <typename T> constexpr size_t maximumSpaceNeededToAlign() {
+  return alignof(T) > alignof(void*) ? sizeof(T) : sizeof(T) + alignof(T) - 1;
```
Comment (mattklein123) [inline source/extensions/filters/common/lua/lua.h]: Should this be:

`return sizeof(T) + (alignof(T) % sizeof(T));` (or something like that?)

Do you have more details on the actual alignment problem that is happening? I'm a little confused as to what the JIT is doing here when user data is allocated.

File: source/extensions/filters/common/lua/lua.h [hunk 2]
```diff
@@ -60,6 +60,28 @@ namespace Lua {
...
+ */
+template <typename T> inline void* allocateLuaUserData(lua_State* state) {
+  auto space = maximumSpaceNeededToAlign<T>();
+  void* mem = lua_newuserdata(state, space);
+  luaL_getmetatable(state, typeid(T).name());
+  ASSERT(lua_istable(state, -1));
+  lua_setmetatable(state, -2);
+
+  // Align the allocated memory.
+  // TODO(dio): should we do #ifdef for skipping aligning the memory?
```
Comment (mattklein123) [inline source/extensions/filters/common/lua/lua.h]: I would skip the ifdef here and below

File: source/extensions/filters/common/lua/lua.h [hunk 3]
```diff
@@ -60,6 +60,28 @@ namespace Lua {
...
+  return std::align(std::alignment_of<T>(), sizeof(mem), mem, space);
```
Comment (mattklein123) [inline source/extensions/filters/common/lua/lua.h]: Is this right? From reading the docs I think it should be `sizeof(T)` and not `sizeof(mem)` ?

File: source/extensions/filters/common/lua/lua.h [hunk 4]
```diff
@@ -119,7 +136,16 @@ template <class T> class BaseLuaObject : protected Logger::Loggable<Logger::Id::
     // manually because the memory is raw and was allocated by Lua.
     to_register.push_back(
         {"__gc", [](lua_State* state) {
-           T* object = static_cast<T*>(luaL_checkudata(state, 1, typeid(T).name()));
+           // TODO(dio): should we do #ifdef for skipping alignment check?
+           auto mem = reinterpret_cast<std::uintptr_t>(luaL_checkudata(state, 1, typeid(T).name()));
+
+           // Check if the allocated memory by Lua was aligned.
```
Comment (mattklein123) [inline source/extensions/filters/common/lua/lua.h]: Can't you just use another std::align call here to make it generic like above?

**Comment by itayd** [comment]
(hey ignore me, just testing something)

/retest

**Pull request closed without merging**